File: app/src/main/java/com/mrit/mesh/transfer/FileTransferManager.kt
package com.mrit.mesh.transfer

import android.util.Log
import com.mrit.mesh.core.MeshID
import com.mrit.mesh.core.MeshPacket
import com.mrit.mesh.core.PacketType
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * FileTransferManager — splits files into 32KB encrypted chunks and reassembles them.
 *
 * Send flow:
 *   1. File is split into [CHUNK_SIZE]-byte chunks
 *   2. Each chunk is wrapped in a chunk payload (see format below)
 *   3. MeshNode encrypts and sends each as a FILE_CHUNK packet
 *
 * Receive flow:
 *   1. FILE_CHUNK packets arrive, are decrypted by MeshNode, passed here
 *   2. Chunks are stored by (transferId, chunkIndex)
 *   3. When all chunks arrive (in any order), the file is reassembled and delivered
 *
 * Chunk payload format (big-endian, binary compatible with iOS FileTransferManager):
 *   [ 16 bytes : transfer ID    ]  random bytes, identifies this file transfer
 *   [  4 bytes : chunk index    ]  0-based, UInt32 big-endian
 *   [  4 bytes : total chunks   ]  UInt32 big-endian
 *   [  4 bytes : total file size]  bytes, UInt32 big-endian
 *   [  2 bytes : filename length]  UInt16 big-endian, 0 for non-first chunks
 *   [  N bytes : filename       ]  UTF-8, only present in chunk 0
 *   [  M bytes : chunk data     ]
 */
class FileTransferManager(
    private val ourId: MeshID,
    private val sendPacket: (MeshPacket, MeshID) -> Unit,
    private val onFileReceived: (fileName: String, data: ByteArray) -> Unit
) {

    companion object {
        private const val TAG        = "FileTransferManager"
        private const val CHUNK_SIZE = 32 * 1024   // 32 KB — well within 65,535-byte packet limit
        private const val HEADER_MIN = 30           // 16+4+4+4+2 bytes before filename
    }

    /** In-progress reassembly: transferId hex → session */
    private val sessions = HashMap<String, ReassemblySession>()
    private val lock     = Any()

    // ── Send ───────────────────────────────────────────────────────────────────

    /**
     * Split [data] into chunks and dispatch as FILE_CHUNK packets to [dest].
     *
     * @param dest     Destination MeshID
     * @param fileName Original filename (e.g. "photo.jpg")
     * @param data     Raw file bytes
     */
    fun send(dest: MeshID, fileName: String, data: ByteArray) {
        val transferId  = java.util.UUID.randomUUID().let {
            ByteBuffer.allocate(16).apply {
                putLong(it.mostSignificantBits)
                putLong(it.leastSignificantBits)
            }.array()
        }
        val totalChunks = (data.size + CHUNK_SIZE - 1) / CHUNK_SIZE

        Log.d(TAG, "Sending '$fileName' (${data.size} bytes) as $totalChunks chunk(s) to ${dest.shortId()}")

        for (i in 0 until totalChunks) {
            val start     = i * CHUNK_SIZE
            val end       = minOf(start + CHUNK_SIZE, data.size)
            val chunkData = data.copyOfRange(start, end)
            val name      = if (i == 0) fileName else ""

            val payload = buildPayload(
                transferId  = transferId,
                chunkIndex  = i,
                totalChunks = totalChunks,
                totalSize   = data.size,
                fileName    = name,
                chunkData   = chunkData
            )

            val packet = MeshPacket(
                type    = PacketType.FILE_CHUNK,
                srcId   = ourId,
                dstId   = dest,
                ttl     = MeshPacket.DEFAULT_TTL,
                payload = payload
            )
            sendPacket(packet, dest)
        }
    }

    // ── Receive ────────────────────────────────────────────────────────────────

    /**
     * Called by MeshNode when a FILE_CHUNK packet arrives and has been decrypted.
     *
     * @param srcId   Sender's MeshID
     * @param payload Decrypted chunk payload bytes
     */
    fun onChunkReceived(srcId: MeshID, payload: ByteArray) {
        val header = parseHeader(payload) ?: run {
            Log.w(TAG, "Malformed FILE_CHUNK payload from ${srcId.shortId()} (${payload.size} bytes)")
            return
        }
        val key = header.transferId.joinToString("") { "%02x".format(it) }

        synchronized(lock) {
            val session = sessions.getOrPut(key) {
                ReassemblySession(
                    transferId  = header.transferId,
                    totalChunks = header.totalChunks,
                    totalSize   = header.totalSize
                )
            }
            if (header.fileName.isNotEmpty()) session.fileName = header.fileName
            session.chunks[header.chunkIndex] = header.chunkData

            Log.d(TAG, "Chunk ${header.chunkIndex+1}/${header.totalChunks} received from ${srcId.shortId()}")

            if (session.chunks.size == session.totalChunks) {
                sessions.remove(key)
                val assembled = session.assemble()
                if (assembled != null) {
                    Log.d(TAG, "File '${session.fileName}' assembled (${assembled.size} bytes)")
                    onFileReceived(session.fileName, assembled)
                }
            }
        }
    }

    // ── Payload codec ──────────────────────────────────────────────────────────

    private fun buildPayload(
        transferId:  ByteArray,
        chunkIndex:  Int,
        totalChunks: Int,
        totalSize:   Int,
        fileName:    String,
        chunkData:   ByteArray
    ): ByteArray {
        val nameBytes = fileName.toByteArray(Charsets.UTF_8)
        val buf = ByteBuffer.allocate(HEADER_MIN + nameBytes.size + chunkData.size)
            .order(ByteOrder.BIG_ENDIAN)
        buf.put(transferId)                      // 16 bytes
        buf.putInt(chunkIndex)                   //  4 bytes
        buf.putInt(totalChunks)                  //  4 bytes
        buf.putInt(totalSize)                    //  4 bytes
        buf.putShort(nameBytes.size.toShort())   //  2 bytes
        buf.put(nameBytes)                       //  N bytes
        buf.put(chunkData)                       //  M bytes
        return buf.array()
    }

    private fun parseHeader(data: ByteArray): ChunkHeader? {
        if (data.size < HEADER_MIN) return null
        val buf = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN)

        val transferId  = ByteArray(16).also { buf.get(it) }
        val chunkIndex  = buf.int
        val totalChunks = buf.int
        val totalSize   = buf.int
        val nameLen     = buf.short.toInt() and 0xFFFF

        if (data.size < HEADER_MIN + nameLen) return null

        val fileName  = if (nameLen > 0) {
            ByteArray(nameLen).also { buf.get(it) }.toString(Charsets.UTF_8)
        } else {
            buf.position(buf.position() + nameLen)
            ""
        }
        val chunkData = ByteArray(buf.remaining()).also { buf.get(it) }

        return ChunkHeader(transferId, chunkIndex, totalChunks, totalSize, fileName, chunkData)
    }

    // ── Data models ────────────────────────────────────────────────────────────

    private data class ChunkHeader(
        val transferId:  ByteArray,
        val chunkIndex:  Int,
        val totalChunks: Int,
        val totalSize:   Int,
        val fileName:    String,
        val chunkData:   ByteArray
    )

    private class ReassemblySession(
        val transferId:  ByteArray,
        val totalChunks: Int,
        val totalSize:   Int
    ) {
        var fileName = ""
        val chunks   = HashMap<Int, ByteArray>()

        /** Combine all chunks in order into the original file bytes. */
        fun assemble(): ByteArray? {
            if (chunks.size != totalChunks) return null
            val result = ByteArray(totalSize)
            var pos = 0
            for (i in 0 until totalChunks) {
                val chunk = chunks[i] ?: return null
                chunk.copyInto(result, pos)
                pos += chunk.size
            }
            return result
        }
    }
}

File: ios/Sources/MritMesh/Mesh/FileTransferManager.swift
import Foundation

/// FileTransferManager (iOS) — splits files into 32KB encrypted chunks and reassembles them.
///
/// Binary-compatible with the Android FileTransferManager.
/// Chunk payload header format (identical on both platforms):
///   [ 16 bytes : transfer ID  ]
///   [  4 bytes : chunk index  ]  big-endian UInt32
///   [  4 bytes : total chunks ]  big-endian UInt32
///   [  4 bytes : total size   ]  big-endian UInt32
///   [  2 bytes : name length  ]  big-endian UInt16  (non-zero only in chunk 0)
///   [  N bytes : filename     ]  UTF-8, only in chunk 0
///   [  M bytes : chunk data   ]
class FileTransferManager {

    private static let chunkSize = 32 * 1024  // 32 KB per chunk

    private let ourId:      MeshID
    private let sendPacket: (MeshPacket, MeshID) -> Void
    private let onFileReceived: (String, Data) -> Void

    /// In-progress reassembly sessions: transferId → session
    private var sessions = [String: ReassemblySession]()
    private let lock     = NSLock()

    init(
        ourId:          MeshID,
        sendPacket:     @escaping (MeshPacket, MeshID) -> Void,
        onFileReceived: @escaping (String, Data) -> Void
    ) {
        self.ourId          = ourId
        self.sendPacket     = sendPacket
        self.onFileReceived = onFileReceived
    }

    // ── Send ───────────────────────────────────────────────────────────────────

    func send(to dest: MeshID, fileName: String, data: Data) {
        let transferId  = Data((0..<16).map { _ in UInt8.random(in: 0...255) })
        let totalChunks = (data.count + Self.chunkSize - 1) / Self.chunkSize

        for i in 0..<totalChunks {
            let start     = i * Self.chunkSize
            let end       = min(start + Self.chunkSize, data.count)
            let chunkData = data[start..<end]
            let name      = (i == 0) ? fileName : ""
            let payload   = buildPayload(
                transferId: transferId, chunkIndex: i, totalChunks: totalChunks,
                totalSize: data.count, fileName: name, chunk: chunkData
            )
            let packet = MeshPacket(
                type: .fileChunk, srcId: ourId, dstId: dest, payload: payload
            )
            sendPacket(packet, dest)
        }
    }

    // ── Receive ────────────────────────────────────────────────────────────────

    func onChunkReceived(srcId: MeshID, payload: Data) {
        guard let header = parseHeader(payload) else { return }

        let key = header.transferId.map { String(format: "%02x", $0) }.joined()
        lock.lock()
        var session = sessions[key] ?? ReassemblySession(
            transferId:  header.transferId,
            totalChunks: header.totalChunks,
            totalSize:   header.totalSize
        )
        if !header.fileName.isEmpty { session.fileName = header.fileName }
        session.chunks[header.chunkIndex] = header.chunkData

        let complete = session.chunks.count == session.totalChunks
        if complete { sessions.removeValue(forKey: key) } else { sessions[key] = session }
        lock.unlock()

        if complete, let assembled = session.assemble() {
            onFileReceived(session.fileName, assembled)
        }
    }

    // ── Payload codec ──────────────────────────────────────────────────────────

    private func buildPayload(
        transferId: Data, chunkIndex: Int, totalChunks: Int,
        totalSize: Int, fileName: String, chunk: Data
    ) -> Data {
        let nameData = Data(fileName.utf8)
        var buf = Data(capacity: 30 + nameData.count + chunk.count)
        buf.append(contentsOf: transferId)                             // 16 bytes
        buf.appendBigEndian(UInt32(chunkIndex))                        //  4 bytes
        buf.appendBigEndian(UInt32(totalChunks))                       //  4 bytes
        buf.appendBigEndian(UInt32(totalSize))                         //  4 bytes
        buf.appendBigEndian(UInt16(nameData.count))                    //  2 bytes
        buf.append(contentsOf: nameData)                               //  N bytes
        buf.append(contentsOf: chunk)                                  //  M bytes
        return buf
    }

    private func parseHeader(_ data: Data) -> ChunkHeader? {
        guard data.count >= 30 else { return nil }
        var offset = 0

        let transferId  = data[offset..<offset+16]; offset += 16
        let chunkIndex  = Int(data.readBigEndianUInt32(at: offset)); offset += 4
        let totalChunks = Int(data.readBigEndianUInt32(at: offset)); offset += 4
        let totalSize   = Int(data.readBigEndianUInt32(at: offset)); offset += 4
        let nameLen     = Int(data.readBigEndianUInt16(at: offset)); offset += 2

        guard data.count >= offset + nameLen else { return nil }
        let fileName = String(data: data[offset..<offset+nameLen], encoding: .utf8) ?? ""
        offset += nameLen

        let chunkData = data[offset...]
        return ChunkHeader(
            transferId:  Data(transferId), chunkIndex: chunkIndex, totalChunks: totalChunks,
            totalSize: totalSize, fileName: fileName, chunkData: Data(chunkData)
        )
    }

    // ── Data models ────────────────────────────────────────────────────────────

    private struct ChunkHeader {
        let transferId:  Data
        let chunkIndex:  Int
        let totalChunks: Int
        let totalSize:   Int
        let fileName:    String
        let chunkData:   Data
    }

    private struct ReassemblySession {
        let transferId:  Data
        let totalChunks: Int
        let totalSize:   Int
        var fileName:    String = ""
        var chunks:      [Int: Data] = [:]

        func assemble() -> Data? {
            guard chunks.count == totalChunks else { return nil }
            var result = Data(capacity: totalSize)
            for i in 0..<totalChunks {
                guard let chunk = chunks[i] else { return nil }
                result.append(chunk)
            }
            return result
        }
    }
}

// ── Data extensions ────────────────────────────────────────────────────────────

private extension Data {
    mutating func appendBigEndian(_ value: UInt32) {
        var v = value.bigEndian
        withUnsafeBytes(of: &v) { append(contentsOf: $0) }
    }
    mutating func appendBigEndian(_ value: UInt16) {
        var v = value.bigEndian
        withUnsafeBytes(of: &v) { append(contentsOf: $0) }
    }
    func readBigEndianUInt32(at offset: Int) -> UInt32 {
        guard count >= offset + 4 else { return 0 }
        return UInt32(self[offset]) << 24 | UInt32(self[offset+1]) << 16
             | UInt32(self[offset+2]) << 8 | UInt32(self[offset+3])
    }
    func readBigEndianUInt16(at offset: Int) -> UInt16 {
        guard count >= offset + 2 else { return 0 }
        return UInt16(self[offset]) << 8 | UInt16(self[offset+1])
    }
}
